include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,3681 @@
+  /// @brief move singleton object into odi map
+  /// @pre the precondition is the object now becomes non-singleton
+  /// @param id an object id
+  void move_singleton_to_odi_map(const obj_id_t &id) {
+    crab::CrabStats::count(domain_name() + ".count.move_object");
+    crab::ScopedCrabStats __st__(domain_name() + ".move_object");
+
+    base_dom_variable_vector_t flds_vec;
+    get_obj_flds_with_ghost(id, flds_vec);
+
+    const odi_domain_product_t *prod_ref = m_odi_map.find(id);
+    if (!prod_ref) {
+      CRAB_ERROR(domain_name(), "::", __func__, ": object ", id,
+                 " value is not found on odi map");
+    }
+    // NOTE: copy the following value is required because
+    // updating the odi map is construcing a new tree based on previous tree
+    odi_domain_product_t res_prod = *prod_ref;
+    object_info_t &obj_info = res_prod.first();
+    // increase reference counting
+    obj_info.refcount_val().increment(id);
+
+    // Two cases here:
+    //  a. if we keep singleton in base, we need to copy its properties
+    //  from base to odi map;
+    //  b. otherwise, although there is no need to update cache,
+    //  we still need to copy from the cache to the summary.
+    // The copy is required because the summary subdomain represents summary
+    // objects which is the abstraction of one or more concrete objects. For
+    // singleton object, there is no need to keep summary object. For
+    // non-singleton, summary object in current state stores properties
+    object_value_t &odi_val = res_prod.second();
+    if (!crab_domain_params_man::get().singletons_in_base()) {
+      // Not necessary to reset cache because there is no ref_store/ref_load
+      // occur. However, we still need to commit cache.
+      reduce_regs_flds_between_cache_and_base(m_base_dom, m_eq_regs_dom,
+                                              odi_val, flds_vec);
+      // copy singleton object into summary
+      odi_val.first() = odi_val.second().first();
+      odi_val.second().second().set_to_top();
+    } else {
+      // TODO: performance may matter if the base dom contains huge dimensions
+      // instead of copy the base domain and perform project, it is better to
+      // have a special project onto operation returns a new domain after
+      // projection.
+      base_abstract_domain_t singleton_base(m_base_dom);
+      m_ghost_var_num_man.project(flds_vec, m_base_dom);
+
+      // copy singleton object into summary
+      odi_val.first() = singleton_base;
+      for (auto &fld : flds_vec) {
+        m_ghost_var_num_man.forget(fld, m_base_dom);
+      }
+    }
+    // update odi map
+    m_odi_map.set(id, res_prod);
+
+    CRAB_LOG("object-move-odi", crab::outs() << "move object: " << id << "\n";
+             m_odi_map.odi_write(crab::outs(), res_prod);
+             crab::outs() << "\n";);
+  }
+
+  /***************** Cache operations *****************/
+  // commit cache contents into summary
```

Comment:
if the object is singleton, you can do a reassign instead of join. I don't know if this case actually happens, but if executing the flush cache intrinsic causes the cache to be flushed and then the cache is updated you would lose precision here